### DISCUSSION TITLE
Unstable custom type causes updates after find

### DISCUSSION THREAD

**Discussion by ladal1** [mikro-orm/mikro-orm#4870] [General]
**Describe the bug**
We have custom type defined which works by parsing the value (for example equivalent operation is appending random number after string when saving to database (`convertToDatabaseValue`) and removing it in `convertToJSValue` function - so the two operations are not symmetrical, only in part of the output). However as the result of `convertToDatabaseValue` is random, it causes mikro orm to immediately update all hydrated entities as soon as `.flush()` is called, is it possible to 

**To Reproduce**
Use type such as this: (define randint however you want)
```
export class SpecialStringType extends Type<string> {
  public convertToDatabaseValue(value?: string): string {
    return `${value}-{randint()}`
  }

  public convertToJSValue(value?: string): string {
    return value.split('-')[0]
  }

  public getColumnType(): string {
    return 'varchar';
  }

  public compareAsType(): string {
    return 'string';
  }
}
```
Use the type as a property and fetch such values.
Flush the EntityManager, you'll see updates for the fetched entities.

**Expected behavior**
Ability to set check on JS values, define custom comparator or simply not updating the first fetched values

**Versions**

| Dependency | Version |
| - | - |
| node | 18.17.1 |
| typescript | 5.2.2 |
| mikro-orm | 6.0.0-dev.91 |
| your-driver | postgresql 6.0.0-dev.91 |

**Comment by B4nan** [comment]
Not sure if I understand the use case here, why would you have a random database value and stable runtime value? If you can describe the use case I can suggest some workarounds.

The change tracking works based on the database values, as that is what we know how to compare, as opposed to the runtime values which can be pretty much anything.

**Reply by B4nan** [reply]
I can imagine how we could get around this in v6 probably, as we ensure change tracking on all properties work regardless of `useDefineForClassFields` option - I think we could completely skip the diffing based on the dirty flag (which gets set when you write to some property on runtime - now it is used for auto-flush detection), which could also significantly speed up flushing for the case where no (or a few) changes are made and a lot of entities are loaded. This way you would not fall into the problem. Need to give it some testing.

edit: but the value would be still considered as changed on a flush if you change some other properties, not sure if that is what you want or not?

**Reply by ladal1** [reply]
My aim would be to be able to compare runtime js values instead of those to be saved to database, as the usage is encryption of the value when saved to database. The value is loaded from the database transformed into readable string and worked with it as it's a normal string.
I was thinking if it was possible to bypass it using virtual column with a specific setter which would change the value?

edit: I would think having a camparator on the custom type would be possible as then it's a responsibility of the developer to make sure their values fit

**Reply by B4nan** [reply]
What if we keep comparing the db values, but allow providing a custom comparator, where you could ignore the random suffix?

**Reply by ladal1** [reply]
If we're able to make function calls in that comparator then it would work fine for us

**Reply by B4nan** [reply]
This should help: ~ff66a99e9f67d2d98b6b9f0252d66c689bb5bee2~ 6d89828fd7770bdd7a5abfb2ef3bb09ad83bb5a7

**Reply by ladal1** [reply]
Looks great, we'll wait for a dev release with it, but based on the code it seems to exactly fit our usage

**Reply by B4nan** [reply]
Hmm the lerna diff check is apparently broken, I just removed it here so it will be out in few minutes.

https://github.com/mikro-orm/mikro-orm/actions/runs/6623912905

**Reply by ladal1** [reply]
Tested the release and it works just fine, however we can't use it yet, as it seems from our tests to have broken parsing of JSON fields when used in query builder. I'll try to figure out if It's something i can fix or expected change

**Reply by ladal1** [reply]
Ah, it's my old friend columnType vs type (first issue I found here 👍 ), rewriting definition of the property from `{ columnType: 'json', type: JsonType, nullable: true }` to `{ type: 'json', nullable: true }` fixed it

**Reply by B4nan** [reply]
That's weird, `type: 'json'` and `type: JsonType` should behave the same. Tried to reproduce in the tests but it seems to be working fine, so repro welcome for that.

**Reply by ladal1** [reply]
Tested it on our project, it seems it's not the `type: JsonType` which is causing the column to not serialize correctly but `columnType: 'json'`
